src/core/sys/posix/pthread.d

```diff
@@ -499,8 +499,11 @@ int pthread_key_delete(pthread_key_t);
 int pthread_mutex_destroy(pthread_mutex_t*);
 int pthread_mutex_init(pthread_mutex_t*, pthread_mutexattr_t*) @trusted;
 int pthread_mutex_lock(pthread_mutex_t*);
+int pthread_mutex_lock(shared(pthread_mutex_t)*);
```

Comment:
There is _zero_ guarantee that operating on something that's `shared` is thread-safe. If anything, it's the opposite. `shared` means that the data is not thread-local. So, proper synchronization is required. To be perfectly correct, basically all extern(C) functions should be have `shared` parameters and return types, because variables in C are not thread-local unless you jump through a bunch of hoops, using stuff that's not actually part of the language.

As it stands, we don't bother marking types on `extern(C)` functions with `shared`, and it's up to the programmer to understand what they're getting into and making sure that that they don't violate the thread-local guarantees that the D compiler provides for D code. If we then decide to start marking `extern(C)` functions with `shared`, the ones that most need to be marked are the ones that _aren't_ thread-safe, not the ones that are.

Deciding to use `shared` on an `extern(C)` function because it's thread-safe, makes no sense at all. If it's thread-safe, then you can get way with passing thread-local data, whereas if it's not, then the data needs to be treated as `shared` by the D code - whether that's by marking up the `extern(C)` function with `shared` or by having the D code just be aware of the semantics and and dealing with them accordingly.

So, if we want to start marking `extern(C)` functions with `shared`, then we should be marking the ones that clearly operate on data that is intended to be shared across threads as `shared`, and whether the function is thread-safe is pretty much irrelevant.